mlir-tensorrt/python/bindings/Compiler/CompilerPyBind.cpp

```diff
@@ -270,7 +275,40 @@ PYBIND11_MODULE(_api, m) {
           py::arg("enabled"),
           py::arg("debug_types") = std::vector<std::string>{},
           py::arg("dump_ir_tree_dir") = py::none(),
-          py::arg("dump_tensorrt_dir") = py::none());
+          py::arg("dump_tensorrt_dir") = py::none())
+
+#ifdef MLIR_TRT_TARGET_TENSORRT
+      .def(
+          "set_tensorrt_translation_metadata_callback",
+          [](PyStableHLOToExecutableOptions &self,
+             std::function<std::string(MlirOperation)> pyCallback) {
+            // Since we're constructing a C callback, our closures must not
+            // capture. We can pass in the Python callback via the userData
+            // argument.
+            auto callback = [](MlirOperation op, MlirStringCallback append,
+                               void *appendCtx, void *userDataVoid) {
+              auto pyCallback =
+                  *static_cast<std::function<std::string(MlirOperation)> *>(
+                      userDataVoid);
+
+              std::string result;
+              try {
+                result = pyCallback(op);
+              } catch (const std::exception &e) {
+                std::cerr << e.what() << std::endl;
+              }
+
+              append(MlirStringRef{result.data(), result.size()}, appendCtx);
+            };
+
+            self.callback = pyCallback;
```

Comment:
Seems like this part is causing segfaults in some cases. Not entirely sure what the cause is, but it could be that the callback is being overwritten by another one and for some reason, something is still pointing to the old address.

Reply:
Interestingly, I cannot reproduce the segault if I run the problematic test(s) individually. It's only when they're run together (even though they are run serially) that the segfault happens.

Reply:
Valgrind output:
```
tests/backend/test_compiler_api.py::TestExecutable::test_incorrect_arguments[args0-kwargs0-Missing argument: b] ==9363== Invalid read of size 8
==9363==    at 0x1BC5B5E1C: std::_Function_base::_M_empty() const (std_function.h:247)
==9363==    by 0x1BC5EE5D4: std::function<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (MlirOperation)>::operator bool() const (std_function.h:573)
==9363==    by 0x1BC5EE541: std::function<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (MlirOperation)>::function(std::function<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (MlirOperation)> const&) (std_function.h:389)
==9363==    by 0x1BC5800F2: pybind11_init__api(pybind11::module_&)::$_3::operator()((anonymous namespace)::PyStableHLOToExecutableOptions&, std::function<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (MlirOperation)>) const::{lambda(MlirOperation, void (*)(MlirStringRef, void*), void*, void*)#1}::operator()(MlirOperation, void (*)(MlirStringRef, void*), void*, void*) const (python/bindings/Compiler/CompilerPyBind.cpp:291)
==9363==    by 0x1BC5800B0: pybind11_init__api(pybind11::module_&)::$_3::operator()((anonymous namespace)::PyStableHLOToExecutableOptions&, std::function<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (MlirOperation)>) const::{lambda(MlirOperation, void (*)(MlirStringRef, void*), void*, void*)#1}::__invoke(MlirOperation, void (*)(MlirStringRef, void*), void*, void*) (python/bindings/Compiler/CompilerPyBind.cpp:288)
```

Reply:
Looks like the problem is that the options instance isn't living long enough:
```
==10578==  Address 0x52588278 is 24 bytes inside a block of size 40 free'd
==10578==    at 0x484C6BF: operator delete(void*) (vg_replace_malloc.c:1131)
==10578==    by 0x1BC57D613: std::default_delete<(anonymous namespace)::PyStableHLOToExecutableOptions>::operator()((anonymous namespace)::PyStableHLOToExecutableOptions*) const (unique_ptr.h:85)
```

Reply:
This is quite strange - at the point where we call `compiler_stablehlo_to_executable`:
```
==8341==    by 0x1C0047526: mtrtCompilerStableHLOToExecutable (compiler/lib/CAPI/Compiler/Compiler.cpp:248)
==8341==    by 0x1BC580B11: pybind11_init__api(pybind11::module_&)::$_4::operator()((anonymous namespace)::PyCompilerClient&, MlirOperation, (anonymous namespace)::PyStableHLOToExecutableOptions&) const (python/bindings/Compiler/CompilerPyBind.cpp:322)
```
the options have already been freed:
```
==8341==  Address 0x518050e8 is 24 bytes inside a block of size 40 free'd
==8341==    at 0x484C6BF: operator delete(void*) (vg_replace_malloc.c:1131)
==8341==    by 0x1BC57D613: std::default_delete<(anonymous namespace)::PyStableHLOToExecutableOptions>::operator()((anonymous namespace)::PyStableHLOToExecutableOptions*) const (unique_ptr.h:85)
==8341==    by 0x1BC57D582: std::unique_ptr<(anonymous namespace)::PyStableHLOToExecutableOptions, std::default_delete<(anonymous namespace)::PyStableHLOToExecutableOptions> >::~unique_ptr() (unique_ptr.h:361)
==8341==    by 0x1BC57D0BB: pybind11::class_<(anonymous namespace)::PyStableHLOToExecutableOptions>::dealloc(pybind11::detail::value_and_holder&) (pybind11.h:1872)
```

This doesn't make sense to me. Shouldn't the arguments to a function live at least for the scope of that function?

Reply:
The problem was that the closure was capturing a pointer by reference instead of value.